File: 2022/july/14-07-2022/construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(self, preorder, inorder):
        if len(inorder) == 0:
            return None

        inorder_index = inorder.index(preorder.pop(0))
        left = self.buildTree(preorder, inorder[:inorder_index])
        right = self.buildTree(preorder, inorder[inorder_index+1:])
        return TreeNode(
            inorder[inorder_index],
            left,
            right,
        )


# tree = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
print(Solution().buildTree([3,9,20,15,7], [9,3,15,20,7]))

File: 2022/july/14-07-2022/construct-binary-tree-from-preorder-and-inorder-traversal.rs
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    fn build(preorder: &mut Vec<i32>, inorder: &mut Vec<i32>, bound: Option<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        if inorder.is_empty() || (bound.is_some() && *inorder.last().unwrap() == bound.unwrap()) {
            return None;
        }
        let mut root = TreeNode::new(preorder.pop().unwrap());
        root.left = Self::build(preorder, inorder, Some(root.val));
        inorder.pop();
        root.right = Self::build(preorder, inorder, bound);
        Some(Rc::new(RefCell::new(root)))
    }


    pub fn build_tree(mut preorder: Vec<i32>, mut inorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        preorder.reverse();
        inorder.reverse();
        Self::build(&mut preorder, &mut inorder, None)
    }
}